File: 12442-forwarding-emails/12442.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <set>
#include <stack>
#include <algorithm>

using namespace std;

struct info {
    int from;
    int to;
    int distance;
    bool visited;
};

int bfs(vector<info>& data, int start)  {
    vector<bool> visited(data.size());
    stack<int> st;
    int dist = 0;
    st.push(start);

    while (!st.empty()) {
        int from = st.top();
        st.pop();
        if (visited[from]) {
            continue;
        }
        visited[from] = true;
        int end = data[from].to;
        /*
        if (data[data[from].to].distance > 0)   {
            dist = data[data[from].to].distance;
            if (data[data[from].to].from != start)   {
                dist++;
            }
            break;
        }
        */
        if (!visited[end])  {
            st.push(end);
            dist++;
        }
    }
    data[start].distance = dist;
    return dist;
/*
    if (data[start].visited)    {
        return 0;
    }
    data[start].visited = true;
    int end = data[start].to;
    return bfs(data, end) + 1;
*/
}

int main()
{
    int t, n, u, v, i, r, m;

    cin >> t;
    for (i = 0; i < t; i++)   {
        cin >> n;
        vector<info> mail(n + 1, info{-1, -1, 0, false});
        while (n-- > 0)  {
            cin >> u >> v;
            mail[u] = info{u, v, 0, false};
        }
        r = 0;
        for (auto& m : mail) {
            if (m.from == -1) {
                continue;
            }
            m.distance = bfs(mail, m.from);
            if (m.distance > r) {
                r = m.distance;
            }
/*            for (auto& d : mail)    {
                if (d.from == -1)   {
                    continue;
                }
                d.visited = false;
            }*/
        }
        for (auto& d : mail)    {
            cout << "from: " << d.from << " to: " << d.to << " distance: " << d.distance << endl;
        }
        set<int> w;
        for (auto& m : mail)    {
            if (m.from == -1)   {
                continue;
            }
            if (m.distance == r)    {
                w.insert(m.from);
            }
        }
        cout << "Case " << i + 1 << ": " << *min_element(w.begin(), w.end()) << endl;
    }
}

File: 12442-forwarding-emails/12442.py
# 12442 Forwarding Emails
# https://onlinejudge.org/external/124/12442.pdf

def bfs(data, start):
    if data[start][3] == True:     # if visited
        return 0
    data[start][3] = True       # mark as visited
    end = data[start][1]
    return bfs(data, end) + 1
    

if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        n = int(input())
        mail = [0] * (n + 1)
        for _ in range(n):
            u, v = [int(x) for x in input().split()]
            mail[u] = [u, v, 0, False]
        # print(mail)
        r = 0
        for m in mail:
            if m == 0:
                continue
            s = m[0]
            m[2] = bfs(mail, s)
            if m[2] > r:
                r = m[2]
            for d in mail:
                if d == 0:
                    continue
                d[3] = False
        w = set()
        for m in mail:
            if m == 0:
                continue
            if m[2] == r:
                w.add(m[0])
        # print(mail, r, min(w))
        print("Case", i + 1, end="")
        print(":", min(w))
